framework/source/class/qx/Promise.js

```diff
@@ -0,0 +1,6547 @@
+    		throw new Error("Promise has already been resolved or rejected");
+    	}
+    	this.__external.complete = true;
+    	return this.__external;
+    },
+    
+    /**
+     * Resolves an external promise
+     */
+    resolve: function(value) {
+    	this.__getPendingExternal().resolve(value);
+    },
+
+    /**
+     * Rejects an external promise
+     */
+    reject: function(reason) {
+    	this.__getPendingExternal().reject(reason);
+    },
+
+    
+    
+    
+    
+    /* *********************************************************************************
+     * 
+     * Utility methods
+     * 
+     */
+    
+    /**
+     * Helper method used to call Promise methods which iterate over an array
+     */
+    _callIterableMethod: function(methodName, args) {
+      args = qx.Promise.__bindArgs(args);
+      return qx.Promise.__wrap(this.__p.then(function(value) {
+        var newP = qx.Promise.Bluebird.resolve(value instanceof qx.data.Array ? value.toArray() : value);
+        return qx.Promise.__wrap(newP[methodName].apply(newP, args));
+      }));
+    },
+    
+    /**
+     * Helper method used to call a Promise method
+     */
+    _callMethod: function(methodName, args) {
+      args = qx.Promise.__bindArgs(args);
+      return qx.Promise.__wrap(this.__p[methodName].apply(this.__p, args));
+    },
+    
+    /**
+     * Returns the actual Promise implementation.
+     * 
+     * Note that Bluebird is the current implementation, and may change without 
+     * notice in the future; if you use this API you accept that this is a private 
+     * implementation detail exposed for debugging or diagnosis purposes only.  For
+     * this reason, the toPromise() method is listed as deprecated starting from the
+     * first release  
+     * @deprecated {6.0} this API method is subject to change
+     */
+    toPromise: function() {
+      return this.__p;
+    }
+  },
+  
+  statics: {
+  	
+    /** Bluebird Promise library; always available */
+    Bluebird: null,
+    
+    /** Native Promise library; only available if the browser supports it */
+    Native: null,
+    
+    /** Promise library, either the Native one or a Polyfill; reliable choice for native Promises */
+    Promise: null,
+    
+    
+    /* *********************************************************************************
+     * 
+     * Promise API methods
+     * 
+     */
+     
+    /**
+     * Returns a Promise object that is resolved with the given value. If the value is a thenable (i.e. 
+     * has a then method), the returned promise will "follow" that thenable, adopting its eventual 
+     * state; otherwise the returned promise will be fulfilled with the value. Generally, if you 
+     * don't know if a value is a promise or not, Promise.resolve(value) it instead and work with 
+     * the return value as a promise.
+     * 
+     * @param value {Object}
+     * @return {qx.Promise}
+     */
+    resolve: function(value) {
```

Comment:
As discussed, `resolve` and 'reject' should take a context parameter to which the Promise is bound, in addition to the value/reason.